Implement POST method

srngch opened the pull request:
## 작업 내용
- closes #14
- POST 구현
- cgi 클래스 분리

hhkim0729 committed:
WIP

srngch committed:
chore: Add python cgi script and Update html form

srngch committed:
chore: Add subject testers

srngch committed:
feat: Implement POST method(#14)

Additional task
- Implement Cgi class to separate logic

hhkim0729 commented on srcs/Cgi.cpp:
```diff
@@ -0,0 +1,130 @@
+		throw std::runtime_error("fail: pipe()");
+	}
+	if (pipe(readPipe) == FT_ERROR) {
+		close(writePipe[0]);
+		close(writePipe[1]);
```
0, 1을 매크로로 지정하는 것이 사용하는 데 헷갈리지 않을 것 같습니다.
READ, WRITE이나 IN, OUT 등..

hhkim0729 commented on srcs/Cgi.cpp:
```diff
@@ -0,0 +1,130 @@
+	close(writePipe[1]);
+
+	memset(readBuf, 0, sizeof(readBuf));
+	while (read(readPipe[0], readBuf, 1024) > 0) {
```
버퍼 사이즈도 매크로로 변경하는 것이 나중에 코드 수정할 때 번거롭지 않을 것 같습니다~!

hhkim0729 commented on srcs/Worker.cpp:
```diff
@@ -135,18 +92,28 @@ ft_bool Worker::work() {
 				if (isCgi(filePath)) {
-					std::string outFile = executeCgiProgram(filePath);
+					Cgi cgi(request_);
+					std::string outFile = cgi.execute();
```
outFile이란 이름이 조금 애매한 것 같긴 하군요..

srngch replied:
result? payload? body?

hhkim0729 replied:
responseBody?

S0YKIM replied:
전 outFile 이나 result 요~

hhkim0729 reviewed (commented):
CGI 실행 코드에 주석이 잘 달려있어서 이해하기 좋았습니다~!

S0YKIM commented on srcs/Cgi.cpp:
```diff
@@ -0,0 +1,130 @@
+
+	env_.push_back("GATEWAY_INTERFACE=CGI/1.1");
+
+	env_.push_back("PATH_INFO=" + request_->getUri()); // TODO: get pathInfo
```
여전히 남아 있는 의문이 subject 에서 full path 를 사용하라는 부분인데 사실상 PATH_INFO 와 PATH_TRANSLATED 를 동일하게 사용하라는 의미가 아닌가 싶습니다 (webroot + uri)

srngch replied:
이 부분은 PATH_INFO에 대해서 더 학습해보고, #23 이슈에서 해결할 예정입니다

S0YKIM commented on srcs/Cgi.cpp:
```diff
@@ -0,0 +1,130 @@
+	// if (request_->getQueryString())
+		env_.push_back("QUERY_STRING="); // TODO: + request_->getQueryString()
+
+	env_.push_back("REMOTE_ADDR=127.0.0.1");
```
이거는 항상 127.0.0.1 로 고정일까요?

srngch replied:
주석 추가 완료!

S0YKIM commented on srcs/Cgi.cpp:
```diff
@@ -0,0 +1,130 @@
+		if (dup2(readPipe[1], STDOUT_FILENO) == FT_ERROR)
+			exit(EXIT_FAILURE);
+		close(readPipe[1]);
+		devNull = open("/dev/null", O_WRONLY);
```
devNull 은 에러 메시지 출력용 파일인가요?

srngch replied:
네!

srngch committed:
refactor: Use macro for pipe in and out fd

macro
- pipe in and out fd
- cgi read buffer size

Additional task
- Add TODO comments

srngch committed:
refactor: Rename variable
